### PR TITLE
ARROW-5911: [Java] Make ListVector and MapVector create reader lazily

### PR DESCRIPTION
**Pull request by liyafan82**
Current implementation creates reader eagerly, which may cause unnecessary resource and time. This issue changes the behavior to lazily create the reader.

This is a follow-up issue for ARROW-5897.

### CODE REVIEW

**Review by BryanCutler — COMMENTED** [review]
Thanks, this looks good! How about we add FixedSizedListVector also to be consistent?

**Comment by liyafan82** [comment]
> Thanks, this looks good! How about we add FixedSizedListVector also to be consistent?

@BryanCutler Sure. Thanks a lot for your kind reminder.
I think there are some other vectors with this problem. Do you think we should change them too (Maybe in another issue, because the change can be substantial, and some refactorings to the class hierarchy are required)?

**Comment by BryanCutler** [comment]
What other vectors did you see with this issue? I did a quick search and only found FixedSizeListVector, the others looked like they create the reader once in the constructor, which is ok

**Comment by liyafan82** [comment]
> What other vectors did you see with this issue? I did a quick search and only found FixedSizeListVector, the others looked like they create the reader once in the constructor, which is ok

@BryanCutler I see. They are indeed different scenarios. Thanks a lot for your clarification. 

So there are some other questions (no for this particular issue). Would you please give your valuable comments?
1. In the other vectors (e.g. IntVector), the reader is created in the constructor but may not be actually used. Do you think these readers should be created lazily?
2. It seems the readers can be extracted to the super class, as a common member?

**Comment by BryanCutler** [comment]
> 1. In the other vectors (e.g. IntVector), the reader is created in the constructor but may not be actually used. Do you think these readers should be created lazily?

Creating these readers isn't expensive. The main reason for doing it in this PR is for readability, since calling `invalidateReader()` is much clearer than making a new reader all over the place. So I don't think it's necessary to change any other readers.

> It seems the readers can be extracted to the super class, as a common member?

I'm not sure if this would help much, but you are welcome to make a PR if you think so

File: java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java [hunk 1]
```diff
@@ -184,9 +183,16 @@ private void setReaderAndWriterIndex() {
 
   @Override
   public UnionFixedSizeListReader getReader() {
+    if (reader == null) {
+      reader = new UnionFixedSizeListReader(this);
+    }
     return reader;
   }
 
+  public void invalidateReader() {
```
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java]: This should be `private`

**Review by BryanCutler — CHANGES_REQUESTED** [review]

**Comment by liyafan82** [comment]
> > 1. In the other vectors (e.g. IntVector), the reader is created in the constructor but may not be actually used. Do you think these readers should be created lazily?
> 
> Creating these readers isn't expensive. The main reason for doing it in this PR is for readability, since calling `invalidateReader()` is much clearer than making a new reader all over the place. So I don't think it's necessary to change any other readers.
> 
> > It seems the readers can be extracted to the super class, as a common member?
> 
> I'm not sure if this would help much, but you are welcome to make a PR if you think so

@BryanCutler I see. Thanks for your valuable feedback.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java (hunk 1 above)
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java]: Revised. Thanks a lot for your careful review.

**Review by BryanCutler — APPROVED** [review]
LGTM, I think the intent is much clearer after this change

**Comment by BryanCutler** [comment]
merged to master, thanks @liyafan82 !

**Pull request closed without merging**